### PR TITLE
Clean executionTaskManager upon detecting an ongoing execution not started by CC.

### PR DESCRIPTION
**Pull request by efeg**
* Fixes the inconsistent state upon rebooted CC instance encountered in in issue https://github.com/linkedin/cruise-control/issues/141. Does not set the `ExecutorState` back to the `STOPPING_EXECUTION`, instead clears the execution task manager state.
* Make Executor thread-safe.
* This patch also fixes the Issue https://github.com/linkedin/cruise-control/issues/138.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 1]
```diff
@@ -113,12 +113,19 @@ public ExecutorState state() {
   }
 
   /**
-   * Kick off the execution.
+   * Pause the load monitor and kick off the execution.
+   *
+   * @param loadMonitor Load monitor.
    */
   public void startExecution(LoadMonitor loadMonitor) {
+    if (loadMonitor != null) {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: If we do not expect the load monitor to be null, can we throw `IllegalArgumentException` and pass a mock object in the unit test?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java [hunk 2]
```diff
@@ -381,8 +381,7 @@ public void resumeLoadMonitorActivity() {
    */
   public void executeProposals(Collection<ExecutionProposal> proposals,
                                Collection<Integer> unthrottledBrokers) {
-    // Pause the load monitor before the proposal execution.
-    _loadMonitor.pauseMetricSampling();
+    // Add execution proposals and start execution.
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java]: Looking at the code. If there are two threads calling executeProposals() with two different set of proposals, CC may execute the both sets of proposals. Maybe we should just let the executor expose a method of `executeProposals()` and synchronize on that.

Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java]: Made the Executor thread-safe.

File: cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricSampleAggregatorState.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 
   MetricSampleAggregatorState(long generation, long windowMs, int completenessCacheSize) {
     super(generation);
-    // Only keep 5 completeness caches.
+    // Only keep as much as _windowStates.size() completeness caches.
```
Comment (becketqin) [inline cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricSampleAggregatorState.java]: Only keep as many as completenessCacheSize entries.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java [hunk 4]
```diff
@@ -321,11 +323,82 @@ private void completeTask(ExecutionTask task) {
...
-  public void clear() {
+  public synchronized void clear() {
     _brokersToSkipConcurrencyCheck.clear();
     _inProgressReplicaMovementsByBrokerId.clear();
     _inProgressPartitions.clear();
     _executionTaskPlanner.clear();
     _executionTaskTracker.clear();
   }
+
+  public synchronized ExecutionState getExecutionState() {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java]: Nit: we have already got many "State". Can we have something like ExecutionTasksSummary here? Also the convention in Kafka, as well as KafkaCruiseControl, is to avoid using "get" in the getter method.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java [hunk 5]
```diff
@@ -321,11 +323,82 @@ private void completeTask(ExecutionTask task) {
...
+    return new ExecutionState(_executionTaskPlanner.remainingReplicaMovements(),
+                              _executionTaskTracker.tasksInState(ExecutionTask.State.IN_PROGRESS),
+                              _executionTaskTracker.tasksInState(ExecutionTask.State.ABORTING),
+                              _executionTaskTracker.tasksInState(ExecutionTask.State.ABORTED),
+                              _executionTaskTracker.tasksInState(ExecutionTask.State.DEAD),
+                              _executionTaskPlanner.remainingDataToMoveInMB());
+  }
+
+  public static class ExecutionState {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java]: Would package private be sufficient?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 6]
```diff
@@ -68,107 +73,140 @@ public Executor(KafkaCruiseControlConfig config, Time time, MetricRegistry dropw
...
-                                      _executionTaskManager.remainingPartitionMovements(),
-                                      _executionTaskManager.inProgressTasks(),
-                                      _executionTaskManager.abortingTasks(),
-                                      _executionTaskManager.abortedTasks(),
-                                      _executionTaskManager.deadTasks(),
-                                      _executionTaskManager.remainingDataToMoveInMB(),
-                                      _finishedDataMovementInMB);
-      default:
-        throw new IllegalStateException("Should never be here!");
+    _stateLock.lock();
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: The locking becomes a little ugly here. Given that the main thing we want to protect is essentially the state query consistency, maybe we can just create an executor state cache object. The state cache will only be updated by the ExecutionRunnable atomically. Hence we don't need to worry about the consistency.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java (hunk 4 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java]: Renamed: `getExecutionState` -> `getExecutionTasksSummary`.
Since this method does not just return an existing object `x`, but building one within this method, it is named `getX` instead of `x`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 7]
```diff
@@ -180,58 +182,94 @@ public void run() {
...
 
     /**
-     * Start the actual execution of the proposals.
+     *
+     *
+     * @param expectedState Current expected state.
+     * @param nextState Next state to be in if the current state is the given expected state.
+     * @param moveRunnable If the current state is the given expected state, runnable to execute after state transition.
+     */
+    private void maybeMove(ExecutorState.State expectedState,
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: I am wondering if this separate method is necessary. The previous logic seem simpler and easier to understand. Is there reason of having this separate method?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 8]
```diff
@@ -180,58 +182,94 @@ public void run() {
...
+        // Wait indefinitely for partition movements to finish.
         waitForExecutionTaskToFinish();
         partitionsToMove = _executionTaskManager.remainingPartitionMovements().size();
         long dataToMove = _executionTaskManager.remainingDataToMoveInMB();
-        _numFinishedPartitionMovements =
-            numTotalPartitionMovements - partitionsToMove - _executionTaskManager.inExecutionTasks().size();
-        _finishedDataMovementInMB = totalDataToMoveInMB - dataToMove;
+        Set<ExecutionTask> inExecutionTasks = _executionTaskManager.inExecutionTasks();
+        _numFinishedPartitionMovements = numTotalPartitionMovements - partitionsToMove - inExecutionTasks.size();
+        _finishedDataMovementInMB = totalDataToMoveInMB - dataToMove - inExecutionDataToMoveInMB(inExecutionTasks);
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: `inExecutionDataToMoveInMB()` may be invoked for N^2 times which could be a waste. Can we just maintain a variable in `ExecutionTaskManager`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 9]
```diff
@@ -180,58 +182,94 @@ public void run() {
...
+    /**
+     * Get the in execution data to move in MB.
+     * @param inExecutionTasks Tasks that are either in progress or aborting.
+     * @return the in execution data to move in MB.
+     */
+    private long inExecutionDataToMoveInMB(Set<ExecutionTask> inExecutionTasks) {
+      return inExecutionTasks.stream().mapToLong(task -> task.proposal().dataToMoveInMB()).sum();
+    }
+
     private void moveReplicas() {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: I did not see the `_executorState` get updated when move the partitions. Am I missing something?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 10]
```diff
@@ -242,27 +280,33 @@ private void moveReplicas() {
...
+                 executionState.remainingPartitionMovements().size(),
+                 executionState.abortingTasks(),
+                 executionState.abortedTasks().size(),
+                 executionState.deadTasks().size(),
+                 executionState.remainingDataToMoveInMB());
       }
     }
 
-    private void moveLeaders() {
+    private void moveLeaderships() {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Orthogonal to this patch and not urgent. We may want to expose the leadership movement to the users as well. This could be a separate patch, though.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java (hunk 9 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Executor state is updated in `maybeMove()`.

Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: But during the movement, the executor state may change, right? i.e. the in-progress tasks, completed tasks, etc.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java (hunk 10 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Thanks for the feedback!
Is this ticket relevant to your request -> https://github.com/linkedin/cruise-control/issues/159, or is this some other functionality?

Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: It is something different. I was thinking about the leadership movement progress.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 11]
```diff
@@ -242,31 +272,40 @@ private void moveReplicas() {
...
       }
       if (_executionTaskManager.inProgressTasks().isEmpty()) {
         LOG.info("Partition movements finished.");
-      } else if (_stopRequested) {
-        LOG.info("Partition movements stopped. {} in progress, {} pending, {} aborted, {} dead.",
-                 _executionTaskManager.inProgressTasks().size(),
-                 _executionTaskManager.remainingPartitionMovements().size(),
-                 _executionTaskManager.abortedTasks().size(),
-                 _executionTaskManager.deadTasks().size());
+      } else if (_stopRequested.get()) {
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: What if the _executionTaskManager does not have more inProgressTasks, but stopRequested == true? It looks we are going to start to move the leaders in that case?

I think we need to updateOngoingExecutionState in the following places:
1. Within waitForExecutionTaskToFinish();
2. Each time after waitForExecutionTaskToFinish() returns;
3. After moveLeaderships() and moveReplicas();

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java (hunk 10 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Yes, that's definitely something we should add. I will create an issue about it for tracking purposes.
